glib/src/subclass/types.rs

```diff
@@ -369,6 +369,31 @@ pub trait ObjectSubclass: Sized + 'static {
     fn with_class(_klass: &Self::Class) -> Self {
         Self::new()
     }
+
+    /// Performs additional instance initialization.
+    ///
+    /// Called just after `with_class()`. At this point the initialization has not completed yet, so
+    /// only a limited set of operations is safe (see `InitializingObject`).
+    fn instance_init(_obj: InitializingObject<Self::Type>) {}
+}
+
+/// An object that is currently being initialized.
+///
+/// Binding crates should use traits for adding methods to this struct. Only methods explicitly safe
+/// to call during `instance_init()` should be added.
+pub struct InitializingObject<T: ObjectType>(T);
+
+impl<T: ObjectType> InitializingObject<T> {
+    /// Returns a reference to the object.
+    ///
+    /// # Safety
+    ///
+    /// The returned object has not been completely initialized at this point. Use of the object
+    /// should be restricted to methods that are explicitly documented to be safe to call during
+    /// `instance_init()`.
+    pub unsafe fn as_ref(&self) -> &T {
```

Comment:
I'm a bit worried here that we provide a normal reference. Anything with that reference is potentially fatal :)

But I guess that's ok, the callers fault if they call an unsafe function!

Reply:
Yeah, that's exactly the motivation for making it `unsafe`. See also `NonNull::as_ref()` etc.

Reply:
Making it return a raw pointer would make it more clear but sure :)